Performance Improvements

joker-at-work opened the pull request:
We've rewritten the parser and the code writing down the logs to be faster and more efficient.

Please see the individual commits' messages for details.

joker-at-work committed:
Add some tests for AsyncTelnet()

Add a validation test and 2 tests showing the performance of the parser.
On my laptop the performance tests take roughly 20s when run one after
the other:

	vspc.tests.test_telnet.AsyncTelnetTest.test_performance_mixed                16.291
	vspc.tests.test_telnet.AsyncTelnetTest.test_performance_lots_of_data         20.047
	vspc.tests.test_telnet.AsyncTelnetTest.test_performance_lots_of_commands     17.741
	vspc.tests.test_telnet.AsyncTelnetTest.test_performance_lots_of_escaped_iac  14.898

joker-at-work committed:
Rewrite AsyncTelnet for performance

We noticed that servers dumping loads of data and no commands into VSPC
would be able to fill up their socket's buffer easily, because VSPC was
not able to handle data fast enough. This came from the parsing and all
the string concatenation stemming from handling a single char at a time.

We rewrote the whole AsyncTelnet handling to make use of string indizes
and keep the creation of new strings (i.e. string concatenation) at a
minimum. This also shows in the performance tests running on my laptop
one after another:

	vspc.tests.test_telnet.AsyncTelnetTest.test_performance_mixed                3.979
	vspc.tests.test_telnet.AsyncTelnetTest.test_performance_lots_of_commands     1.424
	vspc.tests.test_telnet.AsyncTelnetTest.test_performance_lots_of_escaped_iac  0.051
	vspc.tests.test_telnet.AsyncTelnetTest.test_performance_lots_of_data         0.007

joker-at-work committed:
Switch tox to py38 env

We switched to Python 3.8 some time ago, but this still referenced py35.

joker-at-work committed:
Ignore *.swp files

Development with VIM creates those.

joker-at-work committed:
tests: Move _run_async() and support multiple awaitables

We move _run_async() into a utils module, because we will need it in
other tests, too. For that, we also extend it to support passing an
Iterable of awaitables instead of only one awaitable, so we can run
multiple coroutines in parallel.

joker-at-work committed:
tests: Add performance tests for writing to log

This simulates a lot of very small updates coming in from one or
multiple instances at the same time.

joker-at-work committed:
Revert "Added aiofiles to read/write files"

This reverts commit bbff0b26e912fa41b710d3efeffc5fb636534fa5.

Pushing every little tiny update into an own thread with
asyncio.run_in_executor() is just too slow to get anything down to disk.
We have to find a better solution for that and revert the change for
now.

The performance tests for writing improve to < 1s (single) and < 20s
(multiple) from ~8s (single) and ~105s (multiple).

joker-at-work committed:
Some code cleanup

joker-at-work committed:
tests: Add sleep to simulate socket read

When mocking AsyncTelnet.read_some(), we now use a coroutine and add an
async.sleep() into the mix to let other coroutines run. This simulates
reading from a socket and shows more similar behavior to the real world,
with sockets having data come in parallel and thus writes happening in
parallel.

joker-at-work committed:
Handle file writes in background thread

When starting the VspcServer, we also start another thread:
BackgroundWriter. This thread is responsible for synchronously writing
the logs to file. We decouple it from the main thread to not let hanging
file writes stop the handling for telnet messages, as this can prohibit
vMotions (the ESXi needs the ACK for a vMotion in <= 5s to actually
start the vMotion).

Since we rely on another thread to be working properly, we introduce a
check method into the main thread. Finding the BackgroundWriter not
alive, we will exit the whole application and hope for k8s or other
means to restart it.

This also increases performance tests speed to < 12s (multiple) and
keeps the single test roughly the same.

We also add a simple writing test here for simple debugging.

joker-at-work committed:
tests: mocked read() with the provided count

Instead of reading all the data, we read only the provided amount of
data in the mocked `_reader.read()` of `AsyncTelnet`. This ensures the
code to fetch more data is actually used and makes it possible to check
the performance of different chunk sizes.

grandchild commented on vspc/tests/test_telnet.py:
```diff
@@ -0,0 +1,87 @@
+        def read(count):
+            if i[0] < len(data):
+                j = i[0]
+                i[0] = i[0] + count
```
Nit:
```suggestion
                i[0] += count
```

grandchild reviewed (approved):
lgtm

joker-at-work committed:
Drop no longer needed writer queues on disconnect

When a client disconnects, we drop the `deque` object after forcing a
flush of the background writer thread and waiting for the queue to be
empty. Not removing the object would create a memory leak.

joker-at-work committed:
Estimate available data instead of counting

Since we run in two threads, the `deque`s holding the to-be-written data
can change while we iterate over them to count the amount of data
available for the decision if we want to start flushing the data.

Therefore, we explicitly count the sum of data we estimate to be in the
`deque` objects. We say "estimate" here, because the whole counting we
do is not thread-safe. Since we flush the queues every 5s anyways, we do
not need too much accuracy.

grandchild commented on vspc/server.py:
```diff
@@ -81,9 +83,74 @@
+        self._should_stop = True
+        self._writes_available.set()
+
+    def set_writes_available(self, count=0, force=False):
```
I think `count` is a bad name for this parameter because it's very non-descript. How about

```suggestion
    def set_writes_available(self, new_data_len=0, force=False):
```
?

joker-at-work replied:
`new_data_len` sounded to me like it would be the new data length, not that it has to be added hm ... not sure if that's just me. What do you think about `data_len` instead?

grandchild replied:
`data_len` is also fine. the `new` suggested to me that this is the length to be added, from incoming, i.e. "new" data. but data_len is already 90% better, so i don't mind the difference.

joker-at-work replied:
I've named it `data_len`.

grandchild reviewed (changes requested):
otherwise lgtm.
